### DISCUSSION TITLE
8주차 Discussion

### DISCUSSION THREAD

**Discussion by chloe-codes1** [slippStudy/java-network-study#14] [General]
## 📢 책을 읽으면서 궁금한 점들을 자유롭게 적고 대답해 주세요

8주차 진행 방식

> 진행자: 윤혁님, 정규님
> 

- 14.4 ~ 15장 읽기
- Discussion 에 질문 or 인상 깊은 점 2개 적어오기

**Comment by jinwookh** [comment]
1. selector 관련한 라인 기술블로그 글. 예전에 재밌게 읽었는데 참고해도 좋을 것 같습니다.
https://engineering.linecorp.com/ko/blog/do-not-block-the-event-loop-part1#mcetoc_1gdcaies0s

2. Selector 문서에는 selector의 키가 thread-safe하지 않다고 하는 반면, SelectionKey 문서에는 키가 thread-safe하다고 나와 있다.
어떤 내용이 맞는 걸까?

3. 15장에서 클라이언트에서 select를 사용하지 않아도 될 것 같다.
select는 여러 개의 IO를 기다리고 있을 때 효과적인 것으로 이해했기 때문이다.

**Comment by yangyunhyeuk** [comment]
### 인상 깊은 점
---
1. 소켓 채널을 생성할 경우에는 항상 소캣 채널과 연결된 소켓이 같이 생성되지만, 소켓을 생성하는 경우에는 소켓 채널이 생성되지 않는 점

2. SelectableChannel 클래스 내 isRegistered() 메서드의 경우, 채널이 Selector에 등록되어 있는지 질의하는 메서드인데 등록되어 있다면 true, 아니면 false를 반환한다. 하지만 true가 리턴되더라도 이 채널이 Selector가 이용할 수 있는 상태라는 보장은 없다. 그 이유는 채널과 Selector의 중계자 역할을 하는 SelectionKey를 취소했을 경우 바로 Selector가 해당 SelectionKey를 제거하고 그 SelectionKey와 연관된 채널을 Selector에서 등록 해제 되는 것이 아니기 때문이다. (458p. 참고)

**Comment by kangbada0728** [comment]
1. 멀티스레드를 포함한 기존 블로킹 IO 의 문제점을 해결하는 방법들 중 JNI 를 이용하여 select, poll 시스템 콜을 사용하는 방법은 효율성은 좋지만, 개발 난이도가 높다고 하였다. 요즘에는 JNI 를 많이 이용할까? 현재 개발하고 있는 서버들을 보면 속도를 요구하는 곳에서는 C++ 을 쓰던데...

**Comment by chloe-codes1** [comment]
### 인상 깊은 점

- Selector 는 세 가지 중 한 가지 selection method를 이용해서 controller 역할을 수행하는데, select()와 select(long timeout)은 blocking인데 selectNow()는 non-locking인 것이 인상깊었다
- SelectionKey 를 잘 관리하는 방법에 대해 알아봐야겠다
  - key 집합 관리, key 등록, 채널 닫을 때 키 해제 등등..

**Comment by LenKIM** [comment]
- Netty 에서 자바 NIO 가 어떻게 사용되고 있을까?
- 비블록킹을 구현하기 위해 여러가지 기법을 사용하는데, 한번 더 언급해봤으면 좋겠다.

**Reply by LenKIM** [reply]
스프링 MVC 는 Java NIO 를 활용하는 건가? 어떻게 동작하는거지?
